jumpscale/sals/process/__init__.py

```diff
@@ -84,595 +141,833 @@ def execute(
 
     Returns:
-        [list(int)] -- list of pids
+        list of int: List of the processes IDs.
     """
-    excludes = excludes or []
     res = []
-    for process in psutil.process_iter():
-        try:
-            cmdline = process.cmdline()
-        except psutil.NoSuchProcess:
-            cmdline = None
-        except psutil.AccessDenied:
-            cmdline = None
-        if cmdline:
-            name = " ".join(cmdline)
+    for process in psutil.process_iter(attrs=["cmdline"]):
+        if process.info["cmdline"]:
+            cmdline = " ".join(process.info["cmdline"])
             for r in regex_list:
-                if name.strip() != "" and re.match(r, name):
+                if re.match(r, cmdline):
                     res.append(process.pid)
     return res
 
 
-def check_start(cmd, filterstr, nrinstances=1, retry=1):
-    """Run command and check if it is started based on filterstr
-
-    Arguments:
-        cmd {str} -- command to be executed
-        filterstr {str} -- filter string
+def check_start(cmd, filterstr, n_instances=1, retry=1, timeout=2, delay=0.5):
+    """Run command (possibly multiple times) and check if it is started based on filterstr
 
-    Keyword Arguments:
-        instances (int) -- number of needed instances (default: {1})
-        retry (int) -- number of retries (default: {1})
+    Args:
+        cmd (str or list of str): Command to be executed.
+        filterstr (str): Filter string. will match against against Process.name(), Process.exe() and Process.cmdline()
+        n_instances (int, optional): Number of needed instances. Defaults to 1.
+        retry (int, optional): Number of retries to execute the command and check. Defaults to 1.
+        timout (int, optional): how long the function should wait for process to finish (first or parnet process started with the command)
+        delay (int, optional): how long the function should delay the checking process after the process finished or timeout exceeded (in case the first process started another one).
 
     Raises:
-        j.exceptions.RuntimeError: will be raised if we didn't reach number of required instances
+        j.exceptions.Runtime: will be raised if we didn't reach number of required instances.
     """
-    found = get_filtered_pids(filterstr)
     for i in range(retry):
-        if len(found) == nrinstances:
-            return
-        # print "START:%s"%cmd
-        execute(cmd)
-        time.sleep(1)
-    found = get_filtered_pids(filterstr)
-    if len(found) != nrinstances:
-        raise j.exceptions.RuntimeError(
-            "could not start %s, found %s nr of instances. Needed %s." % (cmd, len(found), nrinstances)
-        )
-
+        if isinstance(cmd, str):
+            args = shlex.split(cmd)
+        proc = psutil.Popen(args, close_fds=True, stdout=subprocess.PIPE, stdin=subprocess.PIPE, stderr=subprocess.PIPE)
+        j.logger.debug(f"attempt no: {i}")
```

Comment:
IMHO, these inner debug logs like this can be removed after making sure everything works :laughing: